### ISSUE TITLE
My data can be used for Uncertainty Evaluation?How can I use the data collected from my own phone to automatically determine the best threshold?

### ISSUE THREAD

**Issue by smart4654154** [BayesRays/BayesRays#15]
I have read Bayes Rays ([bayesrays.github.io](http://bayesrays.github.io/)) and thank you very much for your work.

In paper 5.2, there is a sentence that goes, "If one selects the best possible threshold value for each scene out of ten equally spaced ones, BayesRays performs Nerfbusters in both PSNR and coverage."

In paper 5.3, there is a sentence that goes, Since our uncertainty field only needs to be calculated once.We suggest that this threshold can serve as real-time user control in interactive NeRF setups like Nerfstudio

Now I want to use the data collected from my own phone to automatically determine the best threshold.

I want to select the best possible threshold value for my scene out of ten equally spacious ones The metric I have chosen is PSNR

My question is:

1. Can I use my own data to complete this idea?

2. I am a beginner, can you provide some code suggestions?

3. Can only Light Field (LF) dataset and ScanNet scenes be used for Uncertainty Evaluation? Can I use the data collected by my own phone?

![image](https://github.com/BayesRays/BayesRays/assets/152783898/ec44470a-a6ff-4d70-b442-1fe3a63c3158)
Thank you very much (I am a beginner and English is poor, and my question may be a simple one. Thank you very much for your help. This paragraph is written using translation software. If there are any discrepancies, please give me. Thank you very much for your help.)

**Comment by lilygoli** [comment]
Hi!

Thanks for reaching out. Yes, you can do this on your own data. 
First of all after running COLMAP on your data you should be able to train a Nerfacto model using the `ns-train nerfacto` command. Then you can extract the uncertainty of your scene using `ns-uncertainty --load-config={PATH_TO_CONFIG} --output-path={PATH_TO_UNCERTAINTY} --lod 8`. After that you can do either of two things to evaluate your model. 
First you can use the interactive viewer through `ns-viewer-u` and the adjustable "threshold" in the Bayes' Rays viewer to adjust and find the best filtering threshold qualitatively.
 For quantitative PSNR evaluation of 10 different thresholds between 0 and 1, you can use this command: `ns-eval-u --load-config {PATH_TO_CONFIG} --output-path {PATH_TO_METRICS} --unc_path {PATH_TO_UNCERTAINTY} --filter-out` and I think the code should work fine on any dataset, using the nerfstudio version 0.3.3 or 0.3.1 that was tested to work with Bayes' Rays. What this command is doing is running the `get_average_filtered_image_metrics` function in `scripts/eval_uncertainty.py` file, that just runs the PSNR evaluation for all ten different thresholds separately. 
I hope this helps. Let me know if you have further questions.

**Comment by smart4654154** [comment]
Unfortunately, I followed your suggestion but encountered an error, which seems to be related to the content of camera_ray_bundle and the get_outputs_for_camera_ray_bundle function(in \nerfstudio\models\base_model.py).Is it my reproduction process error? Or is this a bug?

This is my reproduction process:
conda create --name nerfstudio -y python=3.8
conda activate nerfstudio
python -m pip install --upgrade pip
pip uninstall torch torchvision functorch tinycudann
pip install torch==2.1.2+cu118 torchvision==0.16.2+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
conda install -c "nvidia/label/cuda-11.8.0" cuda-toolkit
pip install ninja git+https://github.com/NVlabs/tiny-cuda-nn/#subdirectory=bindings/torch
pip install -e git+https://github.com/nerfstudio-project/nerfstudio.git@e2e5637d05fc281a28abe7c4f9c86a93e130a085#egg=nerfstudio
Download BayesRays code
pip install -e
ns-process-data {images, video} --data {DATA_PATH} --output-dir {PROCESSED_DATA_DIR}（I am using nerfstudio's sample data: kitchen）
ns-train nerfacto --data {PROCESSED_DATA_DIR}
ns-uncertainty --load-config {PATH_TO_CONFIG} --output-path {PATH_TO_UNCERTAINTY/npy.npy} --lod 8
ns-eval-u --load-config {PATH_TO_CONFIG} --output-path {PATH_TO_METRICS} --unc_path {PATH_TO_UNCERTAINTY} --filter-out

When I run ns-eval-u --load-config {PATH_TO_CONFIG} --output-path {PATH_TO_METRICS} --unc_path {PATH_TO_UNCERTAINTY} --filter-out, an error occurs
![image](https://github.com/BayesRays/BayesRays/assets/152783898/ad52b95a-a3cd-465b-aff7-ce7a4d5cfadc)

I changed the height, width=camera_ray_bundle. shape to:
Width=camera ray bundle. width
Height=camera ray bundle. Height

Then another error occurred: 
 
![image](https://github.com/BayesRays/BayesRays/assets/152783898/5e6bbde1-51e3-4bb0-b819-400bec959d40)

![image](https://github.com/BayesRays/BayesRays/assets/152783898/2d66705c-ddc1-4927-90cf-d3c24eb4981d)

I printed the content of camera_ray_bundle:
Cameras (camera_to_worlds=tensor([[[ 0.0336, -0.8333,  0.5518, -0.3371],
         [ 0.0587,  0.5528,  0.8313,  0.4765],
         [-0.9977,  0.0045,  0.0675,  0.1048]]], device='cuda:0'), fx=tensor([[386.0386]], device='cuda:0'), fy=tensor([[386.0386]], device='cuda:0'), cx=tensor([[248.8499]],
device='cuda:0'), cy=tensor([[186.1730]], device='cuda:0'), width=tensor([[497]], device='cuda:0'), height=tensor([[369]], device='cuda:0'), distortion_params=tensor([[0., 0., 0., 0., 
0., 0.]], device='cuda:0'), camera_type=tensor([[1]], device='cuda:0'), times=None, metadata={})
I find that the content of camera_ray_bundle has no content 'direction'

Continue reading get_outputs_for_camera_ray_bundle function
This code requires camera ray bundle. origins, camera ray bundle. get_row_major_sliced_ray_bundle, but camera ray bundle does not have these content.

**Comment by lilygoli** [comment]
The first error that you are getting is due to a discrepancy in the `FixedIndicesEvalDataloader` in `main` branch and `nerfbusters-changes` branches of nerfstudio, where one loades `cameras` and the other `camera_ray_bundle` ([here](https://github.com/nerfstudio-project/nerfstudio/blob/da2f833e3b615aa3f5a021cc775edd5296171818/nerfstudio/data/utils/dataloaders.py#L203) and [here](https://github.com/nerfstudio-project/nerfstudio/blob/b03e9beccdee2807bea71783f42b71fad5f2992e/nerfstudio/data/utils/dataloaders.py#L190)). This piece of evaluation code was only tested for the `nerfbusters-changes` branch. 
However, this should be easily fixable using the `camera.generate_rays` function (code and documentation reference [here](https://github.com/nerfstudio-project/nerfstudio/blob/da2f833e3b615aa3f5a021cc775edd5296171818/nerfstudio/cameras/cameras.py#L320)) to get the ray bundles from cameras as needed and pass it along to the metrics functions.
Hope this helps! Otherwise you could try running the code using the `nerfbusters-changes` branch as the installation of your nerfstudio,  which is very similar to the `main` branch but only has some differences in the data loaders.

**Comment by smart4654154** [comment]
Thank you very much for your reply. I have chosen to install nerfbusters-changes. As a beginner, without a tutorial, it is difficult for me to successfully install. I am using Win10, can you provide a similar line of code?
![image](https://github.com/BayesRays/BayesRays/assets/152783898/c7ca35e2-22a8-4112-b25a-13636eee6191)
In addition, I found that the pyproject.toml of nerfbusters-changes requires "torch>=1.12.1,<2.0.0", but the pyproject.toml of BayesRays requires>=2.0.1. What should I do?
![image](https://github.com/BayesRays/BayesRays/assets/152783898/d8277f2f-7c4d-4ae1-9b98-c8cc9ca897af)
![image](https://github.com/BayesRays/BayesRays/assets/152783898/a1a4e366-663c-42dd-bb6f-363a874bd0c5)

**Comment by lilygoli** [comment]
You should be able to install nerfbusters-changes branch with a modified pip install:
`pip install git+https://github.com/nerfstudio-project/nerfstudio.git@nerfbusters-changes`
You can use the nerfbusters-changes requirements. Refer to [here](https://github.com/ethanweber/nerfbusters) for the requirement installations (steps 1 and 2)
